File: server/server/controller.go
package server

import (
	"crypto/rand"
	"encoding/hex"
	"errors"
	"fmt"
	"sync"

	"github.com/labstack/echo/v4"
)

const maxSessions = 10_000 // simple protection

type Controller struct {
	sessions map[string]*Session
	lock     sync.Mutex
}

func NewController() *Controller {
	return &Controller{sessions: make(map[string]*Session)}
}

// CreateSession initie une nouvelle session de courses,
// avec la liste donnée.
func (ct *Controller) CreateSession(c echo.Context) error {
	var list ShopList
	if err := c.Bind(&list); err != nil {
		return err
	}
	id, err := ct.createSession(list)
	if err != nil {
		return err
	}
	out := struct {
		SessionID string `json:"sessionID"`
	}{SessionID: id}
	return c.JSON(200, out)
}

// renvoie l'id créé
func (ct *Controller) createSession(list ShopList) (string, error) {
	ct.lock.Lock()
	defer ct.lock.Unlock()

	if len(ct.sessions) >= maxSessions {
		return "", errors.New("internal error: maximum number of session reached")
	}
	id := randString()
	_, has := ct.sessions[id]
	for has {
		id = randString()
		_, has = ct.sessions[id]
	}
	s := Session{id: id, list: list}

	ct.sessions[id] = &s

	return id, nil
}

func (ct *Controller) GetSession(c echo.Context) error {
	id := c.QueryParam("sessionID")

	ct.lock.Lock()
	defer ct.lock.Unlock()
	l, has := ct.sessions[id]
	if !has {
		return fmt.Errorf("La session %s est invalide ou terminée.", id)
	}

	return c.JSON(200, l.list)
}

type update struct {
	Checked bool `json:"checked"`
	ID      int  `json:"id"`
}

func (ct *Controller) UpdateSession(c echo.Context) error {
	var args update
	if err := c.Bind(&args); err != nil {
		return err
	}

	id := c.QueryParam("sessionID")
	l, err := ct.updateSession(args, id)
	if err != nil {
		return err
	}

	return c.JSON(200, l.list)
}

func (ct *Controller) updateSession(in update, sessionID string) (*Session, error) {
	ct.lock.Lock()
	defer ct.lock.Unlock()

	l, has := ct.sessions[sessionID]
	if !has {
		return nil, fmt.Errorf("La session %s est invalide ou terminée.", sessionID)
	}

	for i, v := range l.list {
		if v.Id == in.ID {
			l.list[i].Checked = in.Checked
		}
	}
	return l, nil
}

func randString() string {
	var buffer [20]byte
	rand.Read(buffer[:])
	return hex.EncodeToString(buffer[:])
}

File: server_node/controller/controller.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Controller = void 0;
const maxSessions = 10000; // simple protection
class Controller {
    constructor() {
        this.sessions = new Map();
        /**
           CreateSession initie une nouvelle session de courses,
           avec la liste donnée.
           */
        this.CreateSession = (req, res) => {
            const list = req.body;
            const id = this.createSession(list);
            res.json({ sessionID: id });
        };
        this.GetSession = (req, res) => {
            const id = (req.query["sessionID"] || "");
            const session = this.sessions.get(id);
            if (session === undefined) {
                throw `La session <${id}> est invalide ou terminée.`;
            }
            res.json(session.list);
        };
        this.UpdateSession = (req, res) => {
            const args = req.body;
            const id = (req.query["sessionID"] || "");
            const l = this.updateSession(args, id);
            res.json(l.list);
        };
    }
    // renvoie l'id créé
    createSession(list) {
        if (this.sessions.size >= maxSessions) {
            throw "internal error: maximum number of session reached";
        }
        let id = randString();
        while (this.sessions.has(id)) {
            id = randString();
        }
        const s = { id: id, list: list };
        this.sessions.set(id, s);
        console.log(`Creating session ${id} with ${list.length} ingredients`);
        return id;
    }
    updateSession(args, sessionID) {
        const l = this.sessions.get(sessionID);
        if (l === undefined) {
            throw `La session <${sessionID}> est invalide ou terminée.`;
        }
        l.list.forEach((v, i) => {
            if (v.id == args.id) {
                l.list[i].checked = args.checked;
            }
        });
        return l;
    }
}
exports.Controller = Controller;
function randString() {
    const L = 16;
    let result = "";
    const characters = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
    const charactersLength = characters.length;
    let counter = 0;
    while (counter < L) {
        result += characters.charAt(Math.floor(Math.random() * charactersLength));
        counter += 1;
    }
    return result;
}
